File: iOS/CMIICollector/GestureClassifier.swift
//
//  GestureClassifier.swift
//  On-device port of tools/gesture_parse.py (Stroke + classify), so gestures.csv
//  matches the offline pipeline column-for-column:
//    wall_ms,kernel_down,kernel_up,type,n_fingers,x0,y0,x1,y1,dur_ms,disp,path_len,mean_vel,max_vel
//
//  A "stroke" spans from the first finger down until the last finger lifts.
//  Thresholds are in POINTS/points-per-second here (getevent used raw device
//  units); calibrate per device — see README.
//
import Foundation
import CoreGraphics

/// Bumped whenever a threshold or a classification rule changes.
///
/// It goes into session.json so a `type` column can be interpreted later.
/// Without it, two sessions recorded a month apart carry the same word "swipe"
/// against different rules and nothing in the data says so — and the rules have
/// already changed twice this month.
///
///  1  original: longPress 500, moveDist 10, flingVel 420; hold shown at the
///     platform's own long-press timeout (iOS 0.35s, Android system setting).
///  2  hold pinned to 500ms on both platforms.
let gestureSpecVersion = 2

struct GestureThresholds {
    var longPressMs: Double = 500
    var moveDist: Double = 10        // points; getevent used 40 raw units
    // Calibrated on iPad Pro 12.9" from session 0824_1043 (4 cued swipes, 4 cued
    // drags by one person). Measured mean_vel: drags 228-365 pt/s, swipes 453-799.
    // The old 600 sat INSIDE the swipe class and misread the two slower swipes as
    // scroll. 420 is the midpoint of the gap, with ~60-90 pt/s margin either side.
    // Thin evidence (n=4 per class, one participant) - revisit with more people.
    // Note max_vel separates even better here (drags 826-1400, swipes 2471-3892);
    // worth evaluating as a rule change, but that must be done on BOTH platforms
    // together since gesture_parse.py mirrors this logic.
    var flingVel: Double = 420       // points/s
    var pinchDelta: Double = 20      // points
    var rotateDeg: Double = 15
    static let `default` = GestureThresholds()

    /// For session.json, so the label can be reinterpreted offline.
    var asDictionary: [String: Any] {
        ["spec_version": gestureSpecVersion,
         "units": "points",
         "density": 1.0,
         "long_press_ms": longPressMs,
         "move_dist": moveDist,
         "fling_vel": flingVel,
         "pinch_delta": pinchDelta,
         "rotate_deg": rotateDeg]
    }
}

struct GestureRecord {
    var wallMs: Int
    var kernelDown: Double
    var kernelUp: Double
    var type: String
    var nFingers: Int
    var x0: Int, y0: Int, x1: Int, y1: Int
    var durMs: Double
    var disp: Double
    var pathLen: Double
    var meanVel: Double
    var maxVel: Double

    static let header = "wall_ms,kernel_down,kernel_up,type,n_fingers,x0,y0,x1,y1,dur_ms,disp,path_len,mean_vel,max_vel"

    var csvRow: String {
        "\(wallMs),\(r6(kernelDown)),\(r6(kernelUp)),\(type),\(nFingers),"
        + "\(x0),\(y0),\(x1),\(y1),\(r1(durMs)),\(r1(disp)),\(r1(pathLen)),\(r1(meanVel)),\(r1(maxVel))"
    }
    private func r1(_ v: Double) -> String { String(format: "%.1f", v) }
    private func r6(_ v: Double) -> String { String(format: "%.6f", v) }
}

private struct Finger {
    var x0: Double, y0: Double
    var x: Double, y: Double
    var path: Double = 0
    var pts: [(Double, Double, Double)] = []   // x, y, kernel-ts
}

/// Feeds on (id, phase, x, y, kernelTs, wallMs); emits a GestureRecord when the
/// last finger of a stroke lifts. Mirrors iter_strokes/Stroke in gesture_parse.py.
final class StrokeAssembler {
    var thr: GestureThresholds
    var onGesture: ((GestureRecord) -> Void)?

    private var active = Set<ObjectIdentifier>()
    private var data: [ObjectIdentifier: Finger] = [:]
    private var order: [ObjectIdentifier] = []      // first-seen order (stable finger[0], [1])
    private var maxFingers = 0
    private var downKts = 0.0
    private var upKts = 0.0
    private var wall0 = 0

    init(thr: GestureThresholds = .default) { self.thr = thr }

    func began(_ id: ObjectIdentifier, _ x: Double, _ y: Double, _ kts: Double, _ wall: Int) {
        if active.isEmpty {                          // new stroke starts
            data.removeAll(); order.removeAll()
            maxFingers = 0; downKts = kts; upKts = kts; wall0 = wall
        }
        active.insert(id)
        if data[id] == nil { order.append(id) }
        data[id] = Finger(x0: x, y0: y, x: x, y: y, pts: [(x, y, kts)])
        maxFingers = max(maxFingers, data.count)
    }

    func moved(_ id: ObjectIdentifier, _ x: Double, _ y: Double, _ kts: Double) {
        guard var f = data[id] else { return }
        f.path += hypot(x - f.x, y - f.y)
        f.x = x; f.y = y; f.pts.append((x, y, kts))
        data[id] = f
    }

    func ended(_ id: ObjectIdentifier, _ x: Double, _ y: Double, _ kts: Double) {
        upKts = kts
        if var f = data[id] { f.x = x; f.y = y; data[id] = f }
        active.remove(id)
        if active.isEmpty { finalizeStroke() }
    }

    private func finalizeStroke() {
        let fingers = order.compactMap { data[$0] }
        guard let p = fingers.max(by: { $0.path < $1.path }) else { return }
        let disp = hypot(p.x - p.x0, p.y - p.y0)
        let durMs = (upKts - downKts) * 1000.0
        let meanV = durMs > 0 ? p.path / (durMs / 1000.0) : 0
        var maxV = 0.0
        for i in 1..<max(p.pts.count, 1) where p.pts.count > 1 {
            let dt = p.pts[i].2 - p.pts[i-1].2
            if dt > 0 {
                let d = hypot(p.pts[i].0 - p.pts[i-1].0, p.pts[i].1 - p.pts[i-1].1)
                maxV = max(maxV, d / dt)
            }
        }
        let type = classify(nFingers: maxFingers, disp: disp, durMs: durMs,
                            meanV: meanV, fingers: fingers)
        onGesture?(GestureRecord(
            wallMs: wall0, kernelDown: downKts, kernelUp: upKts, type: type,
            nFingers: maxFingers,
            x0: Int(p.x0.rounded()), y0: Int(p.y0.rounded()),
            x1: Int(p.x.rounded()),  y1: Int(p.y.rounded()),
            durMs: durMs, disp: disp, pathLen: p.path, meanVel: meanV, maxVel: maxV))
        data.removeAll(); order.removeAll()
    }

    private func classify(nFingers nf: Int, disp: Double, durMs: Double,
                          meanV: Double, fingers: [Finger]) -> String {
        if nf >= 2, fingers.count >= 2 {
            let a = fingers[0], b = fingers[1]
            let d0 = hypot(b.x0 - a.x0, b.y0 - a.y0)
            let d1 = hypot(b.x - a.x, b.y - a.y)
            let ang0 = atan2(b.y0 - a.y0, b.x0 - a.x0)
            let ang1 = atan2(b.y - a.y, b.x - a.x)
            let dAng = abs((ang1 - ang0) * 180.0 / .pi)
            if abs(d1 - d0) > thr.pinchDelta { return "pinch" }
            return dAng > thr.rotateDeg ? "rotate" : "multi_tap"
        }
        if disp > thr.moveDist {
            return meanV > thr.flingVel ? "swipe" : "scroll"
        }
        return durMs > thr.longPressMs ? "long_press" : "tap"
    }
}

File: Android/app/src/main/java/com/cmii/collector/GestureClassifier.kt
package com.cmii.collector

import kotlin.math.abs
import kotlin.math.atan2
import kotlin.math.hypot

/**
 * On-device port of tools/gesture_parse.py, and a line-for-line match with the
 * iPad's GestureClassifier.swift.
 *
 * Three implementations of this logic now exist - Python for offline parsing,
 * Swift on the iPad, Kotlin here - and they must agree, because a classifier
 * that quietly disagrees between platforms turns "which tablet was it" into a
 * confound in the labels themselves. Any threshold change has to be made in all
 * three at once.
 *
 * UNITS. iOS reports touches in POINTS; Android reports them in PIXELS. On a 2x
 * device every distance and velocity therefore arrives doubled, and a 264 dp/s
 * drag reads as a 529 dp/s flick against a threshold calibrated in points - both
 * platforms writing "swipe" and "scroll" while meaning different things. The
 * thresholds are scaled by display density instead of scaling the data, so the
 * CSV keeps the digitiser's own resolution and analysis can convert with the
 * density recorded in session.json.
 *
 * Scaling makes the two platforms agree in principle; it does not make the
 * numbers right. They were calibrated on an iPad Pro 12.9" from four swipes and
 * four drags by one person, and the Pixel Tablet is a different size with a
 * different digitiser. Recalibrate from Android strokes before reading anything
 * into the swipe-versus-scroll split here.
 */
data class GestureThresholds(
    val longPressMs: Double = 500.0,
    val moveDist: Double = 10.0,      // pixels, once scaled
    /** Calibrated on iPad Pro 12.9": drags 228-365 pt/s, swipes 453-799, so 420
     *  is the midpoint of the gap. Provisional on Android until re-measured. */
    val flingVel: Double = 420.0,     // pixels/s, once scaled
    val pinchDelta: Double = 20.0,
    val rotateDeg: Double = 15.0
) {
    /**
     * The same thresholds expressed in this display's pixels.
     *
     * Duration and angle are unit-free and stay put; only the three distance and
     * velocity limits move. On a 2x tablet 420 pt/s becomes 840 px/s, which is
     * the same physical hand speed.
     */
    fun scaled(density: Float) = copy(
        moveDist = moveDist * density,
        flingVel = flingVel * density,
        pinchDelta = pinchDelta * density
    )
}

data class GestureRecord(
    val wallMs: Long,
    val kernelDown: Double,
    val kernelUp: Double,
    val type: String,
    val nFingers: Int,
    val x0: Int, val y0: Int, val x1: Int, val y1: Int,
    val durMs: Double,
    val disp: Double,
    val pathLen: Double,
    val meanVel: Double,
    val maxVel: Double
) {
    companion object {
        const val HEADER =
            "wall_ms,kernel_down,kernel_up,type,n_fingers,x0,y0,x1,y1," +
            "dur_ms,disp,path_len,mean_vel,max_vel"
    }

    val csvRow: String
        get() = "$wallMs,${r6(kernelDown)},${r6(kernelUp)},$type,$nFingers," +
                "$x0,$y0,$x1,$y1,${r1(durMs)},${r1(disp)},${r1(pathLen)}," +
                "${r1(meanVel)},${r1(maxVel)}"

    private fun r1(v: Double) = "%.1f".format(v)
    private fun r6(v: Double) = "%.6f".format(v)
}

private class Finger(val x0: Double, val y0: Double) {
    var x = x0
    var y = y0
    var path = 0.0
    val pts = ArrayList<Triple<Double, Double, Double>>()   // x, y, kernel-ts
}

/**
 * A "stroke" spans from the first finger down until the last finger lifts.
 * Mirrors iter_strokes/Stroke in gesture_parse.py.
 */
class StrokeAssembler(private val thr: GestureThresholds = GestureThresholds()) {

    private val active = HashSet<Int>()
    private val data = HashMap<Int, Finger>()
    private val order = ArrayList<Int>()      // first-seen order, so finger[0] is stable
    private var maxFingers = 0
    private var downKts = 0.0
    private var upKts = 0.0
    private var wall0 = 0L

    fun reset() {
        active.clear(); data.clear(); order.clear()
        maxFingers = 0; downKts = 0.0; upKts = 0.0; wall0 = 0
    }

    fun began(id: Int, x: Double, y: Double, kts: Double, wall: Long) {
        if (active.isEmpty()) {                 // a new stroke starts
            data.clear(); order.clear()
            maxFingers = 0; downKts = kts; upKts = kts; wall0 = wall
        }
        active.add(id)
        if (!data.containsKey(id)) order.add(id)
        data[id] = Finger(x, y).also { it.pts.add(Triple(x, y, kts)) }
        maxFingers = maxOf(maxFingers, data.size)
    }

    fun moved(id: Int, x: Double, y: Double, kts: Double) {
        val f = data[id] ?: return
        f.path += hypot(x - f.x, y - f.y)
        f.x = x; f.y = y
        f.pts.add(Triple(x, y, kts))
    }

    /** Returns the finished stroke when the LAST finger lifts, else null. */
    fun ended(id: Int, x: Double, y: Double, kts: Double): GestureRecord? {
        upKts = kts
        data[id]?.let { it.x = x; it.y = y }
        active.remove(id)
        if (active.isNotEmpty()) return null
        return finalizeStroke()
    }

    private fun finalizeStroke(): GestureRecord? {
        val fingers = order.mapNotNull { data[it] }
        val p = fingers.maxByOrNull { it.path } ?: return null
        val disp = hypot(p.x - p.x0, p.y - p.y0)
        val durMs = (upKts - downKts) * 1000.0
        val meanV = if (durMs > 0) p.path / (durMs / 1000.0) else 0.0
        var maxV = 0.0
        for (i in 1 until p.pts.size) {
            val dt = p.pts[i].third - p.pts[i - 1].third
            if (dt > 0) {
                val d = hypot(p.pts[i].first - p.pts[i - 1].first,
                              p.pts[i].second - p.pts[i - 1].second)
                maxV = maxOf(maxV, d / dt)
            }
        }
        val type = classify(maxFingers, disp, durMs, meanV, fingers)
        val rec = GestureRecord(
            wallMs = wall0, kernelDown = downKts, kernelUp = upKts, type = type,
            nFingers = maxFingers,
            x0 = Math.round(p.x0).toInt(), y0 = Math.round(p.y0).toInt(),
            x1 = Math.round(p.x).toInt(), y1 = Math.round(p.y).toInt(),
            durMs = durMs, disp = disp, pathLen = p.path, meanVel = meanV, maxVel = maxV)
        data.clear(); order.clear()
        return rec
    }

    private fun classify(nf: Int, disp: Double, durMs: Double,
                         meanV: Double, fingers: List<Finger>): String {
        if (nf >= 2 && fingers.size >= 2) {
            val a = fingers[0]; val b = fingers[1]
            val d0 = hypot(b.x0 - a.x0, b.y0 - a.y0)
            val d1 = hypot(b.x - a.x, b.y - a.y)
            val ang0 = atan2(b.y0 - a.y0, b.x0 - a.x0)
            val ang1 = atan2(b.y - a.y, b.x - a.x)
            val dAng = abs((ang1 - ang0) * 180.0 / Math.PI)
            if (abs(d1 - d0) > thr.pinchDelta) return "pinch"
            return if (dAng > thr.rotateDeg) "rotate" else "multi_tap"
        }
        if (disp > thr.moveDist) {
            return if (meanV > thr.flingVel) "swipe" else "scroll"
        }
        return if (durMs > thr.longPressMs) "long_press" else "tap"
    }
}
